File: grpc/Q2132_ObjectShippingQuote.java
import GRPC.ShippingQuote;
import GRPC.ShippingQuoteAnswer;
import GRPC.ShippingQuoteData;
import GRPC.SubmitResponse;
import GRPC.TypedAnswer;
import GRPC.TypedData;
import GRPC.TypedJudgeServiceGrpc;
import GRPC.TypedJudgeServiceGrpc.TypedJudgeServiceBlockingStub;
import GRPC.TypedRequest;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import java.util.concurrent.TimeUnit;

public class Q2132_ObjectShippingQuote {
    static final String EXAM_IP        = "36.50.135.242";
    static final int    EXAM_PORT      = 2240;
    static final String STUDENT_CODE   = "B22DCCN863";   // TODO
    static final String QUESTION_ALIAS = "TODO_alias";   // TODO

    public static void main(String[] args) throws InterruptedException {
        ManagedChannel channel = ManagedChannelBuilder
                .forAddress(EXAM_IP, EXAM_PORT).usePlaintext().build();
        try {
            TypedJudgeServiceBlockingStub stub = TypedJudgeServiceGrpc.newBlockingStub(channel);

            TypedData resp = stub.requestTyped(TypedRequest.newBuilder()
                    .setStudentCode(STUDENT_CODE)
                    .setQuestionAlias(QUESTION_ALIAS)
                    .build());

            String requestId = resp.getRequestId();
            ShippingQuoteData data = resp.getShippingQuote();
            double weightKg   = data.getWeightKg();
            int    maxEtaDays = data.getMaxEtaDays();
            System.out.println("requestId=" + requestId + " orderId=" + data.getOrderId()
                    + " weight=" + weightKg + " maxEta=" + maxEtaDays);

            String bestCarrier = null;
            double bestFee     = Double.POSITIVE_INFINITY;
            int    bestEta     = -1;
            double bestReliab  = -1.0;

            for (ShippingQuote q : data.getQuotesList()) {
                if (q.getEtaDays() > maxEtaDays) continue;
                double fee = q.getBaseFee() + weightKg * q.getPerKgFee();
                fee = Math.round(fee * 100.0) / 100.0;
                if (fee < bestFee || (fee == bestFee && q.getReliability() > bestReliab)) {
                    bestCarrier = q.getCarrier();
                    bestFee     = fee;
                    bestEta     = q.getEtaDays();
                    bestReliab  = q.getReliability();
                }
            }
            if (bestCarrier == null) throw new RuntimeException("Không có quote hợp lệ");
            System.out.printf("best: carrier=%s fee=%.2f eta=%d%n", bestCarrier, bestFee, bestEta);

            ShippingQuoteAnswer answer = ShippingQuoteAnswer.newBuilder()
                    .setCarrier(bestCarrier)
                    .setTotalFee(bestFee)
                    .setEtaDays(bestEta)
                    .build();

            SubmitResponse sr = stub.submitTyped(TypedAnswer.newBuilder()
                    .setStudentCode(STUDENT_CODE)
                    .setQuestionAlias(QUESTION_ALIAS)
                    .setRequestId(requestId)
                    .setShippingQuote(answer)
                    .build());

            System.out.println("status=" + sr.getStatus() + ", message=" + sr.getMessage());
        } finally {
            channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        }
    }
}

File: grpc/Q2132_ObjectShippingQuote.py
import math
import grpc

import typed_judge_pb2
import typed_judge_pb2_grpc

EXAM_IP        = "36.50.135.242"
EXAM_PORT      = 2240
STUDENT_CODE   = "B22DCCN863"   # TODO
QUESTION_ALIAS = "TODO_alias"   # TODO


def main():
    with grpc.insecure_channel(f"{EXAM_IP}:{EXAM_PORT}") as channel:
        stub = typed_judge_pb2_grpc.TypedJudgeServiceStub(channel)

        resp = stub.RequestTyped(typed_judge_pb2.TypedRequest(
            student_code=STUDENT_CODE,
            question_alias=QUESTION_ALIAS,
        ))
        request_id  = resp.request_id
        data        = resp.shipping_quote
        weight_kg   = data.weight_kg
        max_eta     = data.max_eta_days
        print(f"requestId={request_id} orderId={data.order_id} "
              f"weight={weight_kg} maxEta={max_eta}")

        best_carrier = None
        best_fee     = math.inf
        best_eta     = -1
        best_reliab  = -1.0

        for q in data.quotes:
            if q.eta_days > max_eta:
                continue
            fee = q.base_fee + weight_kg * q.per_kg_fee
            fee = round(fee, 2)
            if fee < best_fee or (fee == best_fee and q.reliability > best_reliab):
                best_carrier = q.carrier
                best_fee     = fee
                best_eta     = q.eta_days
                best_reliab  = q.reliability

        if best_carrier is None:
            raise RuntimeError("Không có quote hợp lệ")
        print(f"best: carrier={best_carrier} fee={best_fee:.2f} eta={best_eta}")

        answer = typed_judge_pb2.ShippingQuoteAnswer(
            carrier=best_carrier,
            total_fee=best_fee,
            eta_days=best_eta,
        )
        sr = stub.SubmitTyped(typed_judge_pb2.TypedAnswer(
            student_code=STUDENT_CODE,
            question_alias=QUESTION_ALIAS,
            request_id=request_id,
            shipping_quote=answer,
        ))
        print(f"status={sr.status}, message={sr.message}")


if __name__ == "__main__":
    main()
